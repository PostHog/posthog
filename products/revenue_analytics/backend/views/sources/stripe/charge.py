from typing import cast
from collections.abc import Iterable

from posthog.hogql import ast
from posthog.temporal.data_imports.sources.stripe.constants import (
    CHARGE_RESOURCE_NAME as STRIPE_CHARGE_RESOURCE_NAME,
)
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable
from posthog.hogql.database.schema.exchange_rate import (
    EXCHANGE_RATE_DECIMAL_PRECISION,
    convert_currency_call,
)
from products.revenue_analytics.backend.views.sources.helpers import (
    is_zero_decimal_in_stripe,
    currency_aware_divider,
    currency_aware_amount,
)

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_source


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    source = handle.source
    if source is None:
        return

    # Get all schemas for the source, avoid calling `filter` and do the filtering on Python-land
    # to avoid n+1 queries
    schemas = source.schemas.all()
    charge_schema = next((schema for schema in schemas if schema.name == STRIPE_CHARGE_RESOURCE_NAME), None)
    if charge_schema is None:
        return

    charge_schema = cast(ExternalDataSchema, charge_schema)
    if charge_schema.table is None:
        return

    table = cast(DataWarehouseTable, charge_schema.table)
    team = table.team

    prefix = view_prefix_for_source(source)

    query = ast.SelectQuery(
        select=[
            # Base fields to allow insights to work (need `distinct_id` AND `timestamp` fields)
            ast.Alias(alias="id", expr=ast.Field(chain=["id"])),
            ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
            ast.Alias(alias="timestamp", expr=ast.Field(chain=["created_at"])),
            # Useful for cross joins
            ast.Alias(alias="customer_id", expr=ast.Field(chain=["customer_id"])),
            ast.Alias(alias="invoice_id", expr=ast.Field(chain=["invoice_id"])),
            # Empty, but required for the `events` view to work
            ast.Alias(alias="session_id", expr=ast.Constant(value=None)),
            ast.Alias(alias="event_name", expr=ast.Constant(value=None)),
            # Compute the original currency, converting to uppercase to match the currency code in the `exchange_rate` table
            ast.Alias(
                alias="original_currency",
                expr=ast.Call(name="upper", args=[ast.Field(chain=["currency"])]),
            ),
            # Compute the original amount in the original currency
            # by looking at the captured amount, effectively ignoring refunded value
            ast.Alias(
                alias="original_amount",
                expr=ast.Call(
                    name="toDecimal",
                    args=[
                        ast.Field(chain=["amount_captured"]),
                        ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION),
                    ],
                ),
            ),
            # Compute whether the original currency is a zero-decimal currency
            # by comparing it against a list of known zero-decimal currencies
            # in Stripe's API
            ast.Alias(
                alias="enable_currency_aware_divider",
                expr=is_zero_decimal_in_stripe(ast.Field(chain=["original_currency"])),
            ),
            # Compute the amount decimal divider, which is 1 for zero-decimal currencies and 100 for others
            # This is used to convert the original amount to the adjusted amount
            currency_aware_divider(),
            # Compute the adjusted original amount, which is the original amount divided by the amount decimal divider
            currency_aware_amount(),
            # Expose the base/converted currency, which is the base currency from the team's revenue config
            ast.Alias(alias="currency", expr=ast.Constant(value=team.base_currency)),
            # Convert the adjusted original amount to the base currency
            ast.Alias(
                alias="amount",
                expr=convert_currency_call(
                    amount=ast.Field(chain=["currency_aware_amount"]),
                    currency_from=ast.Field(chain=["original_currency"]),
                    currency_to=ast.Field(chain=["currency"]),
                    timestamp=ast.Call(
                        name="_toDate",
                        args=[
                            ast.Call(
                                name="ifNull",
                                args=[
                                    ast.Field(chain=["timestamp"]),
                                    ast.Call(name="toDateTime", args=[ast.Constant(value=0)]),
                                ],
                            ),
                        ],
                    ),
                ),
            ),
        ],
        # Simple query, simply refer to the `stripe_charge` table
        select_from=ast.JoinExpr(table=ast.Field(chain=[table.name])),
        # Only include succeeded charges because they're the ones that represent revenue
        where=ast.CompareOperation(
            left=ast.Field(chain=["status"]),
            right=ast.Constant(value="succeeded"),
            op=ast.CompareOperationOp.Eq,
        ),
    )

    yield BuiltQuery(key=str(table.id), prefix=prefix, query=query)
