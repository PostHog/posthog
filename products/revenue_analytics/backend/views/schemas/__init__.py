from posthog.schema import DatabaseSchemaManagedViewTableKind
from ._definitions import Schema

from .subscription import SCHEMA as SUBSCRIPTION
from .charge import SCHEMA as CHARGE
from .customer import SCHEMA as CUSTOMER
from .product import SCHEMA as PRODUCT
from .revenue_item import SCHEMA as REVENUE_ITEM

SCHEMAS: dict[DatabaseSchemaManagedViewTableKind, Schema] = {
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_SUBSCRIPTION: SUBSCRIPTION,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_REVENUE_ITEM: REVENUE_ITEM,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CHARGE: CHARGE,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CUSTOMER: CUSTOMER,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_PRODUCT: PRODUCT,
}
