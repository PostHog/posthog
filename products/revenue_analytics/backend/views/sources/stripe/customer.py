from typing import cast
from collections.abc import Iterable

from posthog.hogql import ast

from posthog.temporal.data_imports.sources.stripe.constants import (
    CUSTOMER_RESOURCE_NAME as STRIPE_CUSTOMER_RESOURCE_NAME,
    INVOICE_RESOURCE_NAME as STRIPE_INVOICE_RESOURCE_NAME,
)
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable
from products.revenue_analytics.backend.views.sources.helpers import (
    extract_json_string,
    get_cohort_expr,
)

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_source


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    source = handle.source
    if source is None:
        return

    # Get all schemas for the source, avoid calling `filter` and do the filtering on Python-land
    # to avoid n+1 queries
    schemas = source.schemas.all()
    customer_schema = next((schema for schema in schemas if schema.name == STRIPE_CUSTOMER_RESOURCE_NAME), None)
    if customer_schema is None:
        return

    customer_schema = cast(ExternalDataSchema, customer_schema)
    if customer_schema.table is None:
        return

    invoice_schema = next((schema for schema in schemas if schema.name == STRIPE_INVOICE_RESOURCE_NAME), None)
    invoice_table = None
    if invoice_schema is not None:
        invoice_schema = cast(ExternalDataSchema, invoice_schema)
        invoice_table = invoice_schema.table
        if invoice_table is not None:
            invoice_table = cast(DataWarehouseTable, invoice_table)

    table = cast(DataWarehouseTable, customer_schema.table)
    prefix = view_prefix_for_source(source)

    query = ast.SelectQuery(
        select=[
            ast.Alias(alias="id", expr=ast.Field(chain=["outer", "id"])),
            ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
            ast.Alias(alias="timestamp", expr=ast.Field(chain=["created_at"])),
            ast.Alias(alias="name", expr=ast.Field(chain=["name"])),
            ast.Alias(alias="email", expr=ast.Field(chain=["email"])),
            ast.Alias(alias="phone", expr=ast.Field(chain=["phone"])),
            ast.Alias(alias="address", expr=ast.Field(chain=["address"])),
            ast.Alias(alias="metadata", expr=ast.Field(chain=["metadata"])),
            ast.Alias(
                alias="country",
                expr=extract_json_string("address", "country"),
            ),
            ast.Alias(alias="cohort", expr=ast.Constant(value=None)),
            ast.Alias(alias="initial_coupon", expr=ast.Constant(value=None)),
            ast.Alias(alias="initial_coupon_id", expr=ast.Constant(value=None)),
        ],
        select_from=ast.JoinExpr(
            alias="outer",
            table=ast.Field(chain=[table.name]),
        ),
    )

    # If there's an invoice table we can generate the cohort entry
    # by looking at the first invoice for each customer
    if invoice_table is not None:
        cohort_alias: ast.Alias | None = next(
            (alias for alias in query.select if isinstance(alias, ast.Alias) and alias.alias == "cohort"), None
        )
        if cohort_alias is not None:
            cohort_alias.expr = ast.Field(chain=["cohort"])

        initial_coupon_alias: ast.Alias | None = next(
            (alias for alias in query.select if isinstance(alias, ast.Alias) and alias.alias == "initial_coupon"),
            None,
        )
        if initial_coupon_alias is not None:
            initial_coupon_alias.expr = ast.Field(chain=["initial_coupon"])

        initial_coupon_id_alias: ast.Alias | None = next(
            (alias for alias in query.select if isinstance(alias, ast.Alias) and alias.alias == "initial_coupon_id"),
            None,
        )
        if initial_coupon_id_alias is not None:
            initial_coupon_id_alias.expr = ast.Field(chain=["initial_coupon_id"])

        if query.select_from is not None and (
            cohort_alias is not None or initial_coupon_alias is not None or initial_coupon_id_alias is not None
        ):
            query.select_from.next_join = ast.JoinExpr(
                alias="cohort_inner",
                table=ast.SelectQuery(
                    select=[
                        ast.Field(chain=["customer_id"]),
                        ast.Alias(alias="cohort", expr=get_cohort_expr("min(created_at)")),
                        ast.Alias(
                            alias="initial_coupon",
                            expr=ast.Call(
                                name="argMin",
                                args=[
                                    extract_json_string("discount", "coupon", "name"),
                                    ast.Field(chain=["created_at"]),
                                ],
                            ),
                        ),
                        ast.Alias(
                            alias="initial_coupon_id",
                            expr=ast.Call(
                                name="argMin",
                                args=[
                                    extract_json_string("discount", "coupon", "id"),
                                    ast.Field(chain=["created_at"]),
                                ],
                            ),
                        ),
                    ],
                    select_from=ast.JoinExpr(alias="invoice", table=ast.Field(chain=[invoice_table.name])),
                    group_by=[ast.Field(chain=["customer_id"])],
                ),
                join_type="LEFT JOIN",
                constraint=ast.JoinConstraint(
                    constraint_type="ON",
                    expr=ast.CompareOperation(
                        left=ast.Field(chain=["cohort_inner", "customer_id"]),
                        right=ast.Field(chain=["outer", "id"]),
                        op=ast.CompareOperationOp.Eq,
                    ),
                ),
            )

    yield BuiltQuery(key=str(table.id), prefix=prefix, query=query)
