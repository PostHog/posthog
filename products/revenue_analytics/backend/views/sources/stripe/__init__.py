from posthog.schema import DatabaseSchemaManagedViewTableKind
from products.revenue_analytics.backend.views.core import Builder

from .charge import build as charge_builder
from .customer import build as customer_builder
from .product import build as product_builder
from .revenue_item import build as revenue_item_builder
from .subscription import build as subscription_builder

BUILDER: Builder = {
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CHARGE: charge_builder,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CUSTOMER: customer_builder,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_PRODUCT: product_builder,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_REVENUE_ITEM: revenue_item_builder,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_SUBSCRIPTION: subscription_builder,
}
