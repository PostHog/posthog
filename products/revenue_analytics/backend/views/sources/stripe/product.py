from typing import cast
from collections.abc import Iterable

from posthog.hogql import ast

from posthog.temporal.data_imports.sources.stripe.constants import (
    PRODUCT_RESOURCE_NAME as STRIPE_PRODUCT_RESOURCE_NAME,
)
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_source


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    source = handle.source
    if source is None:
        return

    # Get all schemas for the source, avoid calling `filter` and do the filtering on Python-land
    # to avoid n+1 queries
    schemas = source.schemas.all()
    product_schema = next((schema for schema in schemas if schema.name == STRIPE_PRODUCT_RESOURCE_NAME), None)
    if product_schema is None:
        return

    product_schema = cast(ExternalDataSchema, product_schema)
    if product_schema.table is None:
        return

    table = cast(DataWarehouseTable, product_schema.table)
    prefix = view_prefix_for_source(source)

    query = ast.SelectQuery(
        select=[
            ast.Alias(alias="id", expr=ast.Field(chain=["id"])),
            ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
            ast.Alias(alias="name", expr=ast.Field(chain=["name"])),
        ],
        select_from=ast.JoinExpr(table=ast.Field(chain=[table.name])),
    )

    yield BuiltQuery(key=str(table.id), prefix=prefix, query=query)
