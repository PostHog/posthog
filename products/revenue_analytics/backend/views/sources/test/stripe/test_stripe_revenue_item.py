from parameterized import parameterized
from typing import Any

from posthog.hogql import ast
from posthog.hogql.query import execute_hogql_query
from posthog.test.base import snapshot_clickhouse_queries

from posthog.temporal.data_imports.sources.stripe.constants import (
    CHARGE_RESOURCE_NAME,
    CUSTOMER_RESOURCE_NAME,
    INVOICE_RESOURCE_NAME,
    PRODUCT_RESOURCE_NAME,
    SUBSCRIPTION_RESOURCE_NAME,
)

from products.revenue_analytics.backend.views.sources.test.stripe.base import StripeSourceBaseTest
from products.revenue_analytics.backend.views.schemas.revenue_item import SCHEMA as REVENUE_ITEM_SCHEMA
from products.revenue_analytics.backend.views.sources.stripe.revenue_item import (
    build,
    _calculate_months_for_period as calculate_months_for_period,
)


class TestRevenueItemStripeBuilder(StripeSourceBaseTest):
    def setUp(self):
        super().setUp()
        self.setup_stripe_external_data_source()

    def test_build_revenue_item_query_with_all_schemas(self):
        """Test building revenue item query when all required schemas exist."""
        # Setup with all relevant schemas for revenue items
        schemas = [
            CHARGE_RESOURCE_NAME,
            CUSTOMER_RESOURCE_NAME,
            INVOICE_RESOURCE_NAME,
            PRODUCT_RESOURCE_NAME,
            SUBSCRIPTION_RESOURCE_NAME,
        ]
        self.setup_stripe_external_data_source(schemas=schemas)

        queries = list(build(self.stripe_handle))

        # Should build queries for revenue items
        self.assertGreater(len(queries), 0)

        revenue_query = queries[0]
        revenue_table = self.get_stripe_table_by_schema_name(INVOICE_RESOURCE_NAME)

        # Test the query structure
        self.assertQueryContainsFields(revenue_query.query, REVENUE_ITEM_SCHEMA)
        self.assertBuiltQueryStructure(
            revenue_query, str(revenue_table.id), f"stripe.{self.external_data_source.prefix}"
        )

        # Print and snapshot the generated HogQL query
        query_sql = revenue_query.query.to_hogql()
        self.assertQueryMatchesSnapshot(query_sql, replace_all_numbers=True)

    def test_build_revenue_item_query_with_minimal_schemas(self):
        """Test building revenue item query with minimal required schemas."""
        # Setup with minimal schemas that might be required
        schemas = [INVOICE_RESOURCE_NAME]
        self.setup_stripe_external_data_source(schemas=schemas)

        queries = list(build(self.stripe_handle))

        # Print and snapshot for minimal case
        if queries:
            query_sql = queries[0].query.to_hogql()
            self.assertQueryMatchesSnapshot(query_sql, replace_all_numbers=True)

    def test_build_with_subscription_schemas_only(self):
        """Test building revenue item query with subscription-related schemas only."""
        # Setup with subscription schemas
        schemas = [
            SUBSCRIPTION_RESOURCE_NAME,
            PRODUCT_RESOURCE_NAME,
        ]
        self.setup_stripe_external_data_source(schemas=schemas)

        queries = list(build(self.stripe_handle))

        # Print and snapshot for subscription case
        if queries:
            query_sql = queries[0].query.to_hogql()
            self.assertQueryMatchesSnapshot(query_sql, replace_all_numbers=True)

    def test_build_with_no_relevant_schemas(self):
        """Test that build returns empty when no relevant schemas exist."""
        # Setup without any relevant schemas
        self.setup_stripe_external_data_source(schemas=[])

        queries = list(build(self.stripe_handle))

        # Should return no queries
        self.assertEqual(len(queries), 0)

    def test_build_with_no_source(self):
        """Test that build returns empty when source is None."""
        handle = self.create_stripe_handle_without_source()

        queries = list(build(handle))

        # Should return no queries
        self.assertEqual(len(queries), 0)

    def test_revenue_item_query_contains_required_fields(self):
        """Test that the generated query contains all required revenue item fields."""
        # Setup with comprehensive schemas
        schemas = [
            CHARGE_RESOURCE_NAME,
            CUSTOMER_RESOURCE_NAME,
            INVOICE_RESOURCE_NAME,
            PRODUCT_RESOURCE_NAME,
        ]
        self.setup_stripe_external_data_source(schemas=schemas)

        queries = list(build(self.stripe_handle))

        if queries:
            revenue_query = queries[0]
            query_sql = revenue_query.query.to_hogql()

            # Check for specific fields in the query based on the revenue item schema
            self.assertIn("id", query_sql)
            self.assertIn("source_label", query_sql)

            # Check that source_label contains the expected prefix
            expected_prefix = f"stripe.{self.external_data_source.prefix}"
            self.assertIn(f"'{expected_prefix}'", query_sql)

    def test_revenue_item_query_with_currency_conversion(self):
        """Test revenue item query includes currency conversion logic."""
        # Set team base currency to test conversion
        self.set_team_base_currency("EUR")

        schemas = [CHARGE_RESOURCE_NAME, INVOICE_RESOURCE_NAME]
        self.setup_stripe_external_data_source(schemas=schemas)

        queries = list(build(self.stripe_handle))

        if queries:
            query_sql = queries[0].query.to_hogql()

            # Check for currency-related fields/functions
            # The specific implementation may vary
            self.assertIn("currency", query_sql.lower())

            # Print and snapshot for currency conversion case
            self.assertQueryMatchesSnapshot(query_sql, replace_all_numbers=True)

    @parameterized.expand(
        [
            # Same month
            ("2021-01-01", "2021-01-07", 1),
            ("2021-01-01", "2021-01-31", 1),
            # Adjacent months but still 1-month period
            ("2021-01-01", "2021-02-01", 1),
            ("2021-01-15", "2021-02-15", 1),
            # Multi-month periods
            ("2021-01-01", "2021-03-31", 3),
            ("2021-01-01", "2021-06-30", 6),
            ("2021-01-01", "2021-12-31", 12),
            # Cross-year periods
            ("2021-12-01", "2022-01-31", 2),
            ("2021-01-01", "2022-12-31", 24),
            # Edge cases - partial months
            ("2021-01-15", "2021-01-31", 1),
            ("2021-01-01", "2021-01-15", 1),
            ("2021-01-25", "2021-02-10", 1),
            # Leap year
            ("2020-02-01", "2020-02-29", 1),
            ("2020-01-01", "2020-12-31", 12),
            # Ever-increasing periods on limit
            ("2021-01-01", "2021-02-01", 1),
            ("2021-01-01", "2021-02-05", 1),
            ("2021-01-01", "2021-02-10", 1),
            ("2021-01-01", "2021-02-14", 1),
            ("2021-01-01", "2021-02-15", 1),
            ("2021-01-01", "2021-02-16", 2),
            ("2021-01-01", "2021-02-17", 2),
            ("2021-01-01", "2021-02-20", 2),
            ("2021-01-01", "2021-02-25", 2),
            ("2021-01-01", "2021-02-28", 2),
        ]
    )
    def test_calculate_months_for_period_parameterized(self, start_date, end_date, expected_months):
        response: list[list[Any]] = execute_hogql_query(
            ast.SelectQuery(
                select=[
                    calculate_months_for_period(
                        start_timestamp=ast.Call(name="toDateTime", args=[ast.Constant(value=start_date)]),
                        end_timestamp=ast.Call(name="toDateTime", args=[ast.Constant(value=end_date)]),
                    )
                ]
            ),
            self.team,
        ).results

        assert response[0][0] == expected_months

    @snapshot_clickhouse_queries
    def test_calculate_months_query_snapshot(self):
        response: list[list[Any]] = execute_hogql_query(
            ast.SelectQuery(
                select=[
                    calculate_months_for_period(
                        start_timestamp=ast.Call(name="toDateTime", args=[ast.Constant(value="2021-01-01")]),
                        end_timestamp=ast.Call(name="toDateTime", args=[ast.Constant(value="2021-01-07")]),
                    )
                ]
            ),
            self.team,
        ).results

        assert response[0][0] == 1
