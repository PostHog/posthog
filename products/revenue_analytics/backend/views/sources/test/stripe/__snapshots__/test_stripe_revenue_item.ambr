# serializer version: 1
# name: TestRevenueItemStripeBuilder.test_build_revenue_item_query_with_all_schemas
  '''
  SELECT if(greater(period_months, 1), concat(invoice_item_id, '_', toString(month_index)), invoice_item_id) AS id,
         invoice_item_id AS invoice_item_id,
         'stripe.stripe_test' AS source_label,
         addMonths(timestamp, month_index) AS timestamp,
         created_at AS created_at,
         ifNull(notEmpty(subscription_id), false) AS is_recurring,
         product_id AS product_id,
         customer_id AS customer_id,
         invoice.id AS invoice_id,
         subscription_id AS subscription_id,
         NULL AS session_id,
         NULL AS event_name,
         JSONExtractString(discount, 'coupon', 'name') AS coupon,
         JSONExtractString(discount, 'coupon', 'id') AS coupon_id,
         upper(currency) AS original_currency,
         toDecimal(amount_captured, 10) AS original_amount,
         in(original_currency,
            ['BIF', 'CLP', 'DJF', 'GNF', 'JPY', 'KMF', 'KRW', 'MGA', 'PYG', 'RWF', 'UGX', 'VND', 'VUV', 'XAF', 'XOF', 'XPF']) AS enable_currency_aware_divider,
           if(enable_currency_aware_divider, toDecimal(1, 10), toDecimal(100, 10)) AS currency_aware_divider,
           divideDecimal(original_amount, currency_aware_divider) AS currency_aware_amount,
           'USD' AS currency,
           divideDecimal(convertCurrency(original_currency, currency, currency_aware_amount, _toDate(ifNull(timestamp, toDateTime(0)))), toDecimal(period_months, 10)) AS amount
  FROM
    (SELECT id,
            created_at,
            customer_id,
            subscription_id,
            discount,
            arrayJoin(JSONExtractArrayRaw(assumeNotNull(lines.data))) AS data,
            JSONExtractString(data, 'id') AS invoice_item_id,
            JSONExtractString(data, 'amount') AS amount_captured,
            JSONExtractString(data, 'currency') AS currency,
            JSONExtractString(data, 'price', 'product') AS product_id,
            fromUnixTimestamp(JSONExtractUInt(data, 'period', 'start')) AS period_start,
            fromUnixTimestamp(JSONExtractUInt(data, 'period', 'end')) AS period_end,
            greatest(_toInt16(round(divide(dateDiff('day', ifNull(period_start, created_at), ifNull(period_end, created_at)), 30.44))), 1) AS period_months,
            arrayJoin(range(0, period_months)) AS month_index,
            ifNull(period_start, created_at) AS timestamp
     FROM stripe_test_invoice
     WHERE paid) AS invoice
  UNION ALL
  SELECT id AS id,
         id AS invoice_item_id,
         'stripe.stripe_test' AS source_label,
         created_at AS timestamp,
         created_at AS created_at,
         false AS is_recurring,
                  NULL AS product_id,
                  customer_id AS customer_id,
                  invoice_id AS invoice_id,
                  NULL AS subscription_id,
                  NULL AS session_id,
                  NULL AS event_name,
                  NULL AS coupon,
                  NULL AS coupon_id,
                  upper(currency) AS original_currency,
                  toDecimal(amount_captured, 10) AS original_amount,
                  in(original_currency,
                     ['BIF', 'CLP', 'DJF', 'GNF', 'JPY', 'KMF', 'KRW', 'MGA', 'PYG', 'RWF', 'UGX', 'VND', 'VUV', 'XAF', 'XOF', 'XPF']) AS enable_currency_aware_divider,
                    if(enable_currency_aware_divider, toDecimal(1, 10), toDecimal(100, 10)) AS currency_aware_divider,
                    divideDecimal(original_amount, currency_aware_divider) AS currency_aware_amount,
                    'USD' AS currency,
                    convertCurrency(original_currency, currency, currency_aware_amount, _toDate(ifNull(timestamp, toDateTime(0)))) AS amount
  FROM stripe_test_charge
  WHERE and(or(isNull(invoice_id), empty(invoice_id)), equals(status, 'succeeded'))
  '''
# ---
# name: TestRevenueItemStripeBuilder.test_build_revenue_item_query_with_minimal_schemas
  '''
  SELECT if(greater(period_months, 1), concat(invoice_item_id, '_', toString(month_index)), invoice_item_id) AS id,
         invoice_item_id AS invoice_item_id,
         'stripe.stripe_test' AS source_label,
         addMonths(timestamp, month_index) AS timestamp,
         created_at AS created_at,
         ifNull(notEmpty(subscription_id), false) AS is_recurring,
         product_id AS product_id,
         customer_id AS customer_id,
         invoice.id AS invoice_id,
         subscription_id AS subscription_id,
         NULL AS session_id,
         NULL AS event_name,
         JSONExtractString(discount, 'coupon', 'name') AS coupon,
         JSONExtractString(discount, 'coupon', 'id') AS coupon_id,
         upper(currency) AS original_currency,
         toDecimal(amount_captured, 10) AS original_amount,
         in(original_currency,
            ['BIF', 'CLP', 'DJF', 'GNF', 'JPY', 'KMF', 'KRW', 'MGA', 'PYG', 'RWF', 'UGX', 'VND', 'VUV', 'XAF', 'XOF', 'XPF']) AS enable_currency_aware_divider,
           if(enable_currency_aware_divider, toDecimal(1, 10), toDecimal(100, 10)) AS currency_aware_divider,
           divideDecimal(original_amount, currency_aware_divider) AS currency_aware_amount,
           'USD' AS currency,
           divideDecimal(convertCurrency(original_currency, currency, currency_aware_amount, _toDate(ifNull(timestamp, toDateTime(0)))), toDecimal(period_months, 10)) AS amount
  FROM
    (SELECT id,
            created_at,
            customer_id,
            subscription_id,
            discount,
            arrayJoin(JSONExtractArrayRaw(assumeNotNull(lines.data))) AS data,
            JSONExtractString(data, 'id') AS invoice_item_id,
            JSONExtractString(data, 'amount') AS amount_captured,
            JSONExtractString(data, 'currency') AS currency,
            JSONExtractString(data, 'price', 'product') AS product_id,
            fromUnixTimestamp(JSONExtractUInt(data, 'period', 'start')) AS period_start,
            fromUnixTimestamp(JSONExtractUInt(data, 'period', 'end')) AS period_end,
            greatest(_toInt16(round(divide(dateDiff('day', ifNull(period_start, created_at), ifNull(period_end, created_at)), 30.44))), 1) AS period_months,
            arrayJoin(range(0, period_months)) AS month_index,
            ifNull(period_start, created_at) AS timestamp
     FROM stripe_test_invoice
     WHERE paid) AS invoice
  '''
# ---
# name: TestRevenueItemStripeBuilder.test_calculate_months_query_snapshot
  '''
  SELECT greatest(toInt16(round(divide(dateDiff('day', toDateTime('2021-01-01', 'UTC'), toDateTime('2021-01-07', 'UTC')), 30.44))), 1) AS `greatest(_toInt16(round(divide(dateDiff('day', toDateTime('2021-01-01'), toDateTime('2021-01-07')), 30.44))), 1)`
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295,
                     allow_experimental_join_condition=1
  '''
# ---
# name: TestRevenueItemStripeBuilder.test_revenue_item_query_with_currency_conversion
  '''
  SELECT if(greater(period_months, 1), concat(invoice_item_id, '_', toString(month_index)), invoice_item_id) AS id,
         invoice_item_id AS invoice_item_id,
         'stripe.stripe_test' AS source_label,
         addMonths(timestamp, month_index) AS timestamp,
         created_at AS created_at,
         ifNull(notEmpty(subscription_id), false) AS is_recurring,
         product_id AS product_id,
         customer_id AS customer_id,
         invoice.id AS invoice_id,
         subscription_id AS subscription_id,
         NULL AS session_id,
         NULL AS event_name,
         JSONExtractString(discount, 'coupon', 'name') AS coupon,
         JSONExtractString(discount, 'coupon', 'id') AS coupon_id,
         upper(currency) AS original_currency,
         toDecimal(amount_captured, 10) AS original_amount,
         in(original_currency,
            ['BIF', 'CLP', 'DJF', 'GNF', 'JPY', 'KMF', 'KRW', 'MGA', 'PYG', 'RWF', 'UGX', 'VND', 'VUV', 'XAF', 'XOF', 'XPF']) AS enable_currency_aware_divider,
           if(enable_currency_aware_divider, toDecimal(1, 10), toDecimal(100, 10)) AS currency_aware_divider,
           divideDecimal(original_amount, currency_aware_divider) AS currency_aware_amount,
           'EUR' AS currency,
           divideDecimal(convertCurrency(original_currency, currency, currency_aware_amount, _toDate(ifNull(timestamp, toDateTime(0)))), toDecimal(period_months, 10)) AS amount
  FROM
    (SELECT id,
            created_at,
            customer_id,
            subscription_id,
            discount,
            arrayJoin(JSONExtractArrayRaw(assumeNotNull(lines.data))) AS data,
            JSONExtractString(data, 'id') AS invoice_item_id,
            JSONExtractString(data, 'amount') AS amount_captured,
            JSONExtractString(data, 'currency') AS currency,
            JSONExtractString(data, 'price', 'product') AS product_id,
            fromUnixTimestamp(JSONExtractUInt(data, 'period', 'start')) AS period_start,
            fromUnixTimestamp(JSONExtractUInt(data, 'period', 'end')) AS period_end,
            greatest(_toInt16(round(divide(dateDiff('day', ifNull(period_start, created_at), ifNull(period_end, created_at)), 30.44))), 1) AS period_months,
            arrayJoin(range(0, period_months)) AS month_index,
            ifNull(period_start, created_at) AS timestamp
     FROM stripe_test_invoice
     WHERE paid) AS invoice
  UNION ALL
  SELECT id AS id,
         id AS invoice_item_id,
         'stripe.stripe_test' AS source_label,
         created_at AS timestamp,
         created_at AS created_at,
         false AS is_recurring,
                  NULL AS product_id,
                  customer_id AS customer_id,
                  invoice_id AS invoice_id,
                  NULL AS subscription_id,
                  NULL AS session_id,
                  NULL AS event_name,
                  NULL AS coupon,
                  NULL AS coupon_id,
                  upper(currency) AS original_currency,
                  toDecimal(amount_captured, 10) AS original_amount,
                  in(original_currency,
                     ['BIF', 'CLP', 'DJF', 'GNF', 'JPY', 'KMF', 'KRW', 'MGA', 'PYG', 'RWF', 'UGX', 'VND', 'VUV', 'XAF', 'XOF', 'XPF']) AS enable_currency_aware_divider,
                    if(enable_currency_aware_divider, toDecimal(1, 10), toDecimal(100, 10)) AS currency_aware_divider,
                    divideDecimal(original_amount, currency_aware_divider) AS currency_aware_amount,
                    'EUR' AS currency,
                    convertCurrency(original_currency, currency, currency_aware_amount, _toDate(ifNull(timestamp, toDateTime(0)))) AS amount
  FROM stripe_test_charge
  WHERE and(or(isNull(invoice_id), empty(invoice_id)), equals(status, 'succeeded'))
  '''
# ---
